content/blog/build-complex-pipelines-with-multiplexer/index.md

```diff
@@ -0,0 +1,347 @@
+pipe.add_component("ranker", TransformersSimilarityRanker(model="intfloat/simlm-msmarco-reranker", top_k=10))
+pipe.add_component("prompt_builder", PromptBuilder(template=prompt_template))
+pipe.add_component("llm", OpenAIGenerator())
+pipe.add_component("answer_builder", AnswerBuilder())
+
+# Connect the Multiplexer to all the components that need the query
+pipe.connect("multiplexer.value", "query_embedder.text")
+pipe.connect("multiplexer.value", "bm25_retriever.query")
+pipe.connect("multiplexer.value", "ranker.query")
+pipe.connect("multiplexer.value", "prompt_builder.question")
+pipe.connect("multiplexer.value", "answer_builder.query")
+
+pipe.connect("query_embedder", "embedding_retriever.query_embedding")
+pipe.connect("embedding_retriever", "doc_joiner.documents")
+pipe.connect("bm25_retriever", "doc_joiner.documents")
+pipe.connect("doc_joiner", "ranker.documents")
+pipe.connect("ranker", "prompt_builder.documents")
+pipe.connect("prompt_builder", "llm")
+pipe.connect("llm.replies", "answer_builder.replies")
+pipe.connect("llm.meta", "answer_builder.meta")
+pipe.connect("doc_joiner", "answer_builder.documents")
+```
+
+![Hybrid search pipeline with a Multiplexer](hybrid_search_with_multiplexer.jpeg)
+
+With this setup, only the Multiplexer expects an input from the user. This makes the `run()` statement very straightforward once again.
+
+```python
+result = pipe.run({"multiplexer": {"value": "What's the capital of France?"}})
+```
+
+## Closing Loops
+
+> 💡 *Do you want to see the code in action? Check out Tutorial __: [Spot LLM hallucinations with the HallucinationChecker](#)*
+
+When your pipeline loops, there is often one component that needs to receive input from several sources: at first by the user, who sets off the loop, and later from other components, when the loop comes around and needs to start again.
+
+An example of a looping pipeline is a **self-correcting RAG pipeline**. It's like a basic RAG pipeline with a Retriever, a PromptBuilder and a Generator, plus an additional component at the end that checks whether the answer given by the LLM really relates with the content of at least one of the documents that were returned by the Retriever. If the check fails, the reply is likely to be a hallucination, so the pipeline loops back and asks the LLM to try again.
+
+Haystack 2.0 doesn't have such component yet, but we can take inspiration from a similar component that was available for Haystack 1, called [`EntailmentChecker`](https://haystack.deepset.ai/integrations/entailment-checker).
+
+I'm going to create a similar component for Haystack 2.0, which we will call `HallucinationChecker`. If you're interested in the exact implementation of this component, you can check it out own here, but given that custom components are not the focus of this post, I won't describe the process. If you're interested, check out [this guide](https://docs.haystack.deepset.ai/v2.0/docs/custom-components) on the topic.
```

Comment:
Can you add high-level explanation of what this component does? I tried to add something here but feel free to change it
```suggestion
I'm going to create a similar component for Haystack 2.0, which we will call `HallucinationChecker`. This component gets both the `sentence` generated by the LLM and the `documents` used by the LLM to generate the sentence. After looking at `sentence` and `documents`, it tried to infer whether the generated sentence was a contradiction or entailment. If you're interested in the exact implementation of this component, you can check it out here, but given that custom components are not the focus of this post, I won't describe the process. If you're interested in creating your custom component, check out [Creating Custom Components](https://docs.haystack.deepset.ai/v2.0/docs/custom-components) guide.
```

Reply:
```suggestion
I'm going to create a similar component for Haystack 2.0, which we will call `HallucinationChecker`. This component gets both the `statement` generated by the LLM and the `documents` used by the LLM to generate the statement. After looking at `statement` and `documents`, it tries to infer whether the generated statement is supported or contradicted by the documents. If you're interested in the exact implementation of this component, you can check out the [tutorial on `HallucinationChecker`](https://colab.research.google.com/drive/1cbHRSSrkadTv-prhnFiG5d9d9RDPCIqu#scrollTo=d7CiPoKUPwuI), or if you're interested in creating your custom component, check out [Creating Custom Components](https://docs.haystack.deepset.ai/v2.0/docs/custom-components) guide.
```